pkg/loghttp/query.go

```diff
@@ -97,52 +130,81 @@ type Stream struct {
 	Entries []Entry  `json:"values"`
 }
 
-// UnmarshalJSON implements the json.Unmarshaler interface.
-func (q *QueryResponseData) UnmarshalJSON(data []byte) error {
-	unmarshal := struct {
-		Type       ResultType      `json:"resultType"`
-		Result     json.RawMessage `json:"result"`
-		Statistics stats.Result    `json:"stats"`
-	}{}
-
-	err := json.Unmarshal(data, &unmarshal)
-	if err != nil {
-		return err
-	}
-
-	var value ResultValue
-
-	// unmarshal results
-	switch unmarshal.Type {
-	case ResultTypeStream:
-		var s Streams
-		err = json.Unmarshal(unmarshal.Result, &s)
-		value = s
-	case ResultTypeMatrix:
-		var m Matrix
-		err = json.Unmarshal(unmarshal.Result, &m)
-		value = m
-	case ResultTypeVector:
-		var v Vector
-		err = json.Unmarshal(unmarshal.Result, &v)
-		value = v
-	case ResultTypeScalar:
-		var v Scalar
-		err = json.Unmarshal(unmarshal.Result, &v)
-		value = v
-	default:
-		return fmt.Errorf("unknown type: %s", unmarshal.Type)
+func (s *Stream) UnmarshalJSON(data []byte) error {
+	if s.Labels == nil {
+		s.Labels = LabelSet{}
 	}
+	return jsonparser.ObjectEach(data, func(key, value []byte, ty jsonparser.ValueType, _ int) error {
+		switch string(key) {
+		case "stream":
+			if err := s.Labels.UnmarshalJSON(value); err != nil {
+				return err
+			}
+		case "values":
+			if ty == jsonparser.Null {
+				return nil
+			}
+			var parseError error
+			_, err := jsonparser.ArrayEach(value, func(value []byte, ty jsonparser.ValueType, _ int, _ error) {
+				if ty == jsonparser.Null {
+					return
+				}
+				var entry Entry
+				if err := entry.UnmarshalJSON(value); err != nil {
+					parseError = err
+					return
+				}
+				s.Entries = append(s.Entries, entry)
+			})
+			if parseError != nil {
+				return parseError
+			}
+			return err
+		}
+		return nil
+	})
+}
 
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (q *QueryResponseData) UnmarshalJSON(data []byte) error {
+	resultType, err := jsonparser.GetString(data, "resultType")
 	if err != nil {
 		return err
 	}
-
-	q.ResultType = unmarshal.Type
-	q.Result = value
-	q.Statistics = unmarshal.Statistics
-
-	return nil
+	q.ResultType = ResultType(resultType)
+
+	return jsonparser.ObjectEach(data, func(key, value []byte, dataType jsonparser.ValueType, _ int) error {
+		switch string(key) {
+		case "result":
+			switch q.ResultType {
+			case ResultTypeStream:
+				ss := Streams{}
+				if err := ss.UnmarshalJSON(value); err != nil {
+					return err
+				}
+				q.Result = ss
+			case ResultTypeMatrix:
+				var m Matrix
+				err = json.Unmarshal(value, &m)
+				q.Result = m
+			case ResultTypeVector:
+				var v Vector
+				err = json.Unmarshal(value, &v)
+				q.Result = v
+			case ResultTypeScalar:
+				var v Scalar
+				err = json.Unmarshal(value, &v)
+				q.Result = v
```

Comment:
How do we handle `err` in these cases? I can't see where that var is being used

Reply:
Good catch ! Resolved now.